File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/ForEachEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import java.util.ArrayList;
import java.util.List;
import org.opencds.cqf.cql.engine.execution.State;

public class ForEachEvaluator {

    public static Object forEach(Object source, Object element, State state) {
        if (source == null || element == null) {
            return null;
        }

        List<Object> retVal = new ArrayList<>();
        for (Object o : (Iterable<?>) source) {
            retVal.add(state.getEnvironment().resolvePath(o, element.toString()));
        }
        return retVal;
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/ForEachEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic

object ForEachEvaluator {
    @JvmStatic
    fun forEach(source: Any?, element: Any?): Any? {
        if (source == null || element == null) {
            return null
        }

        return (source as Iterable<*>).map { PropertyEvaluator.resolvePath(it, element.toString()) }
    }
}
